nautobot/extras/apps.py

```diff
@@ -31,3 +31,10 @@ def ready(self):
         from nautobot.extras.health_checks import DatabaseBackend
 
         plugin_dir.register(DatabaseBackend)
+
+        # Register built-in SecretsProvider classes
+        from nautobot.extras.secrets.providers import EnvironmentVariableSecretsProvider, TextFileSecretsProvider
+        from nautobot.extras.secrets import register_secrets_provider
+
+        register_secrets_provider(EnvironmentVariableSecretsProvider)
+        register_secrets_provider(TextFileSecretsProvider)
```

Comment:
This feels a little clunky to me. Seeing as these are just classes, and not model objects, AND `SecretsProvider` inherits from `ABC`, we could take advantage of automatic and/or built-in registration by nature of the subclasses being present. See: https://docs.python.org/3/library/abc.html#abc.ABCMeta.register

Or, I am also of the mind that we are more declarative and make this a thing for settings, e.g. 

```python
# nautobot_config.py
SECRETS_PROVIDERS = [
    "nautobot.extras.secrets.providers.EnvironmentVariableSecretsProvider",
    "nautobot.extras.secrets.providers.TextFileSecretsProvider",
]
```

... And then iteratively register them like this, kind of how we do w/ `PLUGINS` (which itself is just extending `INSTALLED_APPS` with some extra conditions), but I think you get my drift.

Reply:
It's a bit clunky for "core" providers, but have to consider the plugin provider pattern. I was trying to follow our existing plugin feature registration pattern for consistency (a plugin can implement custom validators, but has to explicitly register them in `custom_validators:custom_validators`; a plugin can implement custom home page items, but has to explicitly register them in `homepage:layout`; a plugin can implement custom Jobs, but has to explicitly register them in `jobs:jobs`, etc.) It's by no means my favorite pattern but I figured the benefit of self-consistency for plugin authors outweighed the clunkiness of it.

If we're willing to break the pattern, I can see any number of ways this could be done, depending on how "automatic" we want things to be:

- Automatic discovery via `__subclasses__` (`for subclass in SecretsProvider.__subclasses__()...`)
- Implementor-driven registration and discovery via packaging `entry_points` (`for entry_point in iter_entry_points("nautobot.secrets.providers")`) - I'd actually initially implemented this feature this way before I remembered that we, for whatever reason, have avoided this pattern in plugins thus far.
- Implementor-driven registration and discovery via `plugin.secrets:providers` pattern (`secrets_providers = import_object(f"{self.__module__}.{self.secrets_providers}")`, matching other plugin feature registration, as described above)
- Opt-in manual discovery via a separate config setting (`SECRETS_PROVIDERS`, as you proposed above)
- Opt-in manual discovery via a reserved key in `PLUGINS_CONFIG`?
- others...?

Reply:
Thanks. Good to think aloud about these things. I think this current implementation is good, because it relates to how it's already being done (and therefore familiar). We should reserve breaking the pattern for a future release where we can perhaps attempt to give them each the same treatment. 

Thank you!